### ISSUE TITLE
IMPORTANT: must revise `four_sum`

### ISSUE THREAD

**Issue by JeffreySarnoff** [JuliaMath/AccurateArithmetic.jl#18]
Somehow, I mis-derived `four_sum`.

oof, not quite right yet

**Comment by JeffreySarnoff** [comment]
q.v. "Algorithms for triple-word arithmetic"
by Nicolas Fabiano and Jean-Michel Muller and Joris Picot
and refs therein
and quaddouble packages (renormalization for unordered 4-tuples)

One thing I will check out is if it works faster to work on a preordered 4 tuple
using 
```
#    sort 4 values with 5 minmaxs in 3 parallel stages

@inline function min_max(a::T, b::T) where {T<:Real}
    return (b < a ? (b, a) : (a, b))
end

function swapsort(a::T, b::T, c::T, d::T) where {T}

    a, b = min_max(a, b)
    c, d = min_max(c, d)

    a, c = min_max(a, c)
    b, d = min_max(b, d)

    b, c = min_max(b, c)

    return a, b, c, d
end
```

**Comment by JeffreySarnoff** [comment]
@ffevotte ^

**Comment by JeffreySarnoff** [comment]
a much faster (4x) implementation (this is fyi, I am still writing tests).
```
@inline max_min(x,y) = ifelse(abs(y) < abs(x), (x,y) , (y,x))

function maxtomin(a::T, b::T, c::T, d::T) where {T}
    a, b = max_min(a, b)
    c, d = max_min(c, d)
    a, c = max_min(a, c)
    b, d = max_min(b, d)
    b, c = max_min(b, c)
    return a, b, c, d
return

function fast_vecsum_errbranch(x1::T,x2::T,x3::T,x4::T) where {T}
    y = zeros(T, 4)
    j = 1
    # e[1] = x1
    # i = 1
    r, t = two_sum(x1, x2)
    if t !== zero(T)
       y[j] = r
       e = t
       j += 1
    else
       e = r
    end
    # i = 2
    r, t = two_sum(e, x3)
    if t !== zero(T)
       y[j] = r
       e = t
       j += 1
    else
       e = r
    end

    y[j], y[j+1] = two_sum(e, x4)
    return y
end

@inline function fast_quadword(x1::T, x2::T, x3::T, x4::T) where {T}
    a,b,c,d = maxtomin(x1,x2,x3,x4)
    return fast_vecsum_errbranch(a,b,c,d)
end
```

**Comment by JeffreySarnoff** [comment]
The error happened with a few permutations of `a,b,c,d` where one or more args were 0.
  
It appears that adding the `min_max` line fixes it.  More testing is underway.
```
@inline max_min(x, y) = x < y ? (y, x) : (x, y)

function ErrorfreeArithmetic.four_sum(a::T,b::T,c::T,d::T) where {T}
    a, d = max_min(a, d)
    t0, t1 = two_sum(a ,  b)
    t0, t2 = two_sum(t0,  c)
    hi, t3 = two_sum(t0,  d)
    t0, t1 = two_sum(t1, t2)
    midhi, t2 = two_sum(t0, t3)
    midlo, lo = two_sum(t1, t2)
    return hi, midhi, midlo, lo
end
```
I will report back.

**Comment by ffevotte** [comment]
For future reference: this should have been fixed by #21
